db/jdbc/src/main/java/io/helidon/db/jdbc/JdbcStatementGeneric.java

```diff
@@ -0,0 +1,203 @@
+                            interceptorQueryFuture.complete(update);
+                            dmlResultFuture.complete(update);
+                            statement.close();
+                        } finally {
+                            conn.close();
+                        }
+                    }
+                } catch (Exception e) {
+                    if (null != conn) {
+                        try {
+                            // we would not close the connection in the resultSetFuture, so we have to close it here
+                            conn.close();
+                        } catch (SQLException ex) {
+                            LOGGER.log(Level.WARNING, "Failed to close connection", ex);
+                        }
+                    }
+                    resultSetFuture.completeExceptionally(e);
+                }
+            });
+        });
+
+
+        /*
+        Futures:
+        interceptorStatementFuture - completed once we call the statement
+        interceptorQueryFuture - completed once we read all records from a query (or finish a DML) - requires count
+        queryResultFuture - completed once we know this is a result set and we have prepared the DbRowResult
+        dmlResultFuture - completed once we know this is a DML statement
+        exceptionFuture - completes in case of any exception
+        resultSetFuture - completes if this is a result set and has the connection & result set objects
+         */
+        // for DML - everything is finished and done
+        /*
+          For Query
+            Ignored:
+                dmlResultFuture
+            Completed:
+                interceptorStatementFuture
+                resultSetFuture
+            Open:
+                interceptorQueryFuture
+                queryResultFuture
+                exceptionFuture
+         */
+
+        // and now, let's construct the DbRowResult
+        resultSetFuture.thenAccept(rsAndConn -> {
+            DbRowResult<DbRow> dbRowResult = JdbcStatementQuery.processResultSet(interceptorQueryFuture,
+                                                                                 resultSetFuture,
+                                                                                 executorService(),
+                                                                                 dbMapperManager(),
+                                                                                 mapperManager());
+            interceptorQueryFuture.exceptionally(throwable -> {
+                exceptionFuture.complete(throwable);
+                return null;
+            });
+            queryResultFuture.complete(dbRowResult);
+        }).exceptionally(throwable -> {
+            interceptorQueryFuture.completeExceptionally(throwable);
+            queryResultFuture.completeExceptionally(throwable);
+            exceptionFuture.complete(throwable);
+            return null;
+        });
+
+        return new DbResult() {
```

Comment:
Here returned anonymous class may be extracted to a static class or it's own classfile for better readibility.